### PR TITLE
Out of band orchestrator discovery

### PR DESCRIPTION
**Pull request by angyangie**
**What does this pull request do? Explain your changes. (required)**

All changes explained here: https://github.com/livepeer/go-livepeer/issues/641

**Does this pull request close any open issues?**

This one: https://github.com/livepeer/go-livepeer/issues/641

**Checklist:**
- [ ] README and other documentation updated
- [ ] Node runs in OSX and devenv
- [x] All tests in `./test.sh` pass

### CODE REVIEW

File: common/db.go [hunk 1]
```diff
@@ -223,6 +265,45 @@ func (db *DB) LastSeenBlock() (*big.Int, error) {
...
+	if err != nil {
+		glog.Error("db: Unable to get orchestrators updated in the last 24 hours", err)
+		return nil, err
+	}
+	orchs := []*DBOrch{}
+	for rows.Next() {
+		var orch DBOrch
+		var serviceURI string
+		var ethereumAddr string
+		if err := rows.Scan(&serviceURI, &ethereumAddr); err != nil {
```
Comment (j0sh) [inline common/db.go:297]: We can avoid using the serviceURI and ethereumAddr variables here, eg with `rows.Scan(&orch.ServiceURI, &orch.EthereumAddr)`

File: discovery/db_discovery.go [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+	node *core.LivepeerNode
+}
+
+func NewDBOrchestratorPoolCache(node *core.LivepeerNode) *DBOrchestratorPoolCache {
+	if node.Eth == nil {
+		glog.Error("Could not refresh DB list of orchestrators: LivepeerNode nil")
+		return nil
+	}
+
+	ticker := time.NewTicker(1 * time.Hour)
```
Comment (j0sh) [inline discovery/db_discovery.go]: Maybe make `1 * time.Hour` a constant?

Also, what do you think of making this overridable so we can explicitly emit ticks during testing? See a general example of this pattern: https://github.com/livepeer/go-livepeer/pull/657#discussion_r244917469

File: discovery/db_discovery.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+func CacheDBOrchs(node *core.LivepeerNode, orchs []*lpTypes.Transcoder) ([]*common.DBOrch, error) {
+	var dbOrchs []*common.DBOrch
+	if orchs == nil {
+		glog.Error("No new DB orchestrators created: no orchestrators found onchain")
+		return dbOrchs, nil
+	}
+
+	for _, orch := range orchs {
+		dbOrch := EthOrchToDBOrch(orch)
+		if err := node.Database.UpdateOrchs(dbOrch); err != nil {
```
Comment (j0sh) [inline discovery/db_discovery.go]: Just a note for later. At some point we'll probably want to do one of a few things (or even a combination of both):

1. Call `GetOrchestratorInfo` prior to inserting to DB to receive and insert ticket parameters. We need to figure out how to schedule this for large O sets (sequentially? asynchronously? in batches?)
2. Consolidate multiple INSERTs into a single statement, eg `INSERT INTO orchestrators (updatedAt, serviceURI, ethereumAddr) VALUES(datetime(), abc, def), (datetime(), ghi, jkl), (datetime(), mno, pqr)`. We want to minimize the amount of time SQLite spends writing to disk. Unfortunately, it seems we can't use prepared statements for this, so we might need to construct the query at runtime and take special care to escape all the input prior to execution. See https://stackoverflow.com/questions/49156991/sqlite-range-bind-or-column-out-of-range-for-insert-statement

File: common/db.go [hunk 4]
```diff
@@ -223,6 +265,45 @@ func (db *DB) LastSeenBlock() (*big.Int, error) {
 	return big.NewInt(lastSeenBlock), nil
 }
 
+func (db *DB) UpdateOrchs(orch *DBOrch) error {
```
Comment (j0sh) [inline common/db.go]: This function (and the associated prepared statement) only updates a single orchestrator; perhaps it should be named `UpdateOrch` instead?

File: common/db_test.go [hunk 5]
```diff
@@ -118,6 +125,47 @@ func profilesMatch(j1 []ffmpeg.VideoProfile, j2 []ffmpeg.VideoProfile) bool {
...
+	defer dbraw.Close()
+
+	orch := NewStubOrch("127.0.0.1:8936", "0xde0B295669a9FD93d5F28D9Ec85E40f4cb697BAe")
+	err = dbh.UpdateOrchs(orch)
+
+	if err != nil {
+		t.Error("Unexpected error in update orchs: ", err)
+	}
+
+	orchs, err := dbh.SelectOrchs()
```
Comment (j0sh) [inline common/db_test.go:154]: Perhaps also add a test for selecting from an empty table?

File: common/db.go (hunk 4 above)
Comment (j0sh) [inline common/db.go]: Also probably needs an `orch == nil` check here

File: common/db_test.go [hunk 6]
```diff
@@ -118,6 +125,47 @@ func profilesMatch(j1 []ffmpeg.VideoProfile, j2 []ffmpeg.VideoProfile) bool {
 	return true
 }
 
+func TestDBOrchestrators(t *testing.T) {
+	dbh, dbraw, err := TempDB(t)
+	defer dbh.Close()
+	defer dbraw.Close()
+
+	orch := NewStubOrch("127.0.0.1:8936", "0xde0B295669a9FD93d5F28D9Ec85E40f4cb697BAe")
+	err = dbh.UpdateOrchs(orch)
```
Comment (j0sh) [inline common/db_test.go]: Perhaps a test for a `nil` orch since that'll crash `UpdateOrchs` at the moment.

File: discovery/db_discovery.go [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+func CacheDBOrchs(node *core.LivepeerNode, orchs []*lpTypes.Transcoder) ([]*common.DBOrch, error) {
+	var dbOrchs []*common.DBOrch
+	if orchs == nil {
+		glog.Error("No new DB orchestrators created: no orchestrators found onchain")
+		return dbOrchs, nil
+	}
+
+	for _, orch := range orchs {
```
Comment (j0sh) [inline discovery/db_discovery.go:77]: Do we need a `if orch == nil { continue }` check here? Only reason for asking is because EthOchToDBOrch has one, so not sure if there's a case where the `orchs` list could contain a nil entry.

File: common/db_test.go [hunk 8]
```diff
@@ -118,6 +125,47 @@ func profilesMatch(j1 []ffmpeg.VideoProfile, j2 []ffmpeg.VideoProfile) bool {
 	return true
 }
 
+func TestDBOrchestrators(t *testing.T) {
```
Comment (eladmallel) [inline common/db_test.go]: I'd like to suggest breaking this down into more succinct or "surgical" test functions, where we state more clearly the scenario and the expected outcome. IMO taking the approach of smaller functions and the specific naming paradigm nudges us to think more carefully about all the cases we want to cover. See this recent example: https://github.com/livepeer/go-livepeer/pull/655/files#diff-c20584ae20b069ed08dc6769a1c19e77.

In general I like structuring the names of test functions in 3 parts:

- The code under test (usually a function name)
- The scenario
- The expected result

While it's not necessarily how we normally name functions in Go, for the sake of readability I also like to separate these parts with an underscore.

Of course one of the downsides could be repetitive setup code, but that can be mitigated with some solid helper functions.

wdyt?

File: common/db.go [hunk 9]
```diff
@@ -117,8 +135,26 @@ func InitDB(dbPath string) (*DB, error) {
...
+	stmt, err := db.Prepare("SELECT serviceURI, ethereumAddr FROM orchestrators WHERE updatedAt >= datetime('now','-1 day')")
+	if err != nil {
+		glog.Error("Unable to select orchestrators updated in the past 24 hours", err)
+		d.Close()
+		return nil, err
+	}
+	d.selectOrchs = stmt
+
+	// updateOrchestrators statement
+	stmt, err = db.Prepare("INSERT OR REPLACE INTO orchestrators(updatedAt, serviceURI, ethereumAddr) VALUES(datetime(), ?, ?)")
```
Comment (eladmallel) [inline common/db.go]: According to [this tutorial](http://www.sqlitetutorial.net/sqlite-replace-statement/) we can shorten the `INSERT OR REPLACE` statement to simply `REPLACE`. wdyt?

Comment (eladmallel) [reply common/db.go]: Plus, I didn't know about the `REPLACE` feature until just now - thanks for this new learning!

File: common/db_test.go [hunk 10]
```diff
@@ -106,6 +106,13 @@ func TestDBVersion(t *testing.T) {
 	}
 }
 
+func NewStubOrch(serviceURI string, ethereumAddr string) *DBOrch {
```
Comment (eladmallel) [inline common/db_test.go]: Why do we need this function? - seems like we can just use `NewDBOrch`

File: common/db_test.go [hunk 11]
```diff
@@ -118,6 +125,47 @@ func profilesMatch(j1 []ffmpeg.VideoProfile, j2 []ffmpeg.VideoProfile) bool {
...
+	if err != nil {
+		t.Error("Unexpected error in update orchs: ", err)
+	}
+
+	orchs, err := dbh.SelectOrchs()
+	if err != nil || len(orchs) != 1 || orchs[0].ServiceURI != orch.ServiceURI {
+		t.Error("Unexpected error in update orchs: ", err)
+	}
+
+	orchUpdate := NewStubOrch("127.0.0.1:8937", "0xde0B295669a9FD93d5F28D9Ec85E40f4cb697BAe")
```
Comment (eladmallel) [inline common/db_test.go]: Does the ETH address need to be this specific value? - if not I personally like the idea of using random values, primarily because it clarifies to the test reader that we don't really care what value we use here ([see example](https://github.com/livepeer/go-livepeer/pull/655/files#diff-c20584ae20b069ed08dc6769a1c19e77R803)).

wdyt?

File: common/db_test.go [hunk 12]
```diff
@@ -118,6 +125,47 @@ func profilesMatch(j1 []ffmpeg.VideoProfile, j2 []ffmpeg.VideoProfile) bool {
...
+	err = dbh.UpdateOrchs(orchUpdate)
+
+	if err != nil {
```
Comment (eladmallel) [inline common/db_test.go]: Would you like to use the new `testify` library we've been using? - it should allow you to replace all these if blocks with a simple `assert.Nil(err)` ([example here](https://github.com/livepeer/go-livepeer/pull/655/files#diff-c20584ae20b069ed08dc6769a1c19e77R756)). wdyt?

File: discovery/db_discovery.go [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+			glog.Error("Error updating Orchestrator in DB: ", err)
+			return nil, err
+		}
+		dbOrchs = append(dbOrchs, dbOrch)
+	}
+	return dbOrchs, nil
+
+}
+
+func EthOrchToDBOrch(orch *lpTypes.Transcoder) *common.DBOrch {
```
Comment (eladmallel) [inline discovery/db_discovery.go]: Do we need to export this function, or can we make it private?

File: discovery/db_discovery.go [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+	if orch == nil {
+		return nil
+	}
+	return common.NewDBOrch(orch.ServiceURI, orch.Address.String())
+}
+
+func (dbo *DBOrchestratorPoolCache) GetOrchestrators(numOrchestrators int) ([]*net.OrchestratorInfo, error) {
```
Comment (eladmallel) [inline discovery/db_discovery.go:48]: My 2 cents on file layout: I like keeping all methods (i.e. functions with a receiver) close together, usually having a section below all the methods for the non-receiver functions, so it ends up looking something like:

```

func NewSomething(...) {

}

// Something methods

// functions Something uses (and don't have a receiver)
```

wdyt?

File: discovery/db_discovery.go (hunk 3 above)
Comment (eladmallel) [reply discovery/db_discovery.go]: Thinking about how to simplify these concerns, seems to me that the simplest approach would be:

- Update the DB with `TicketParam`s using an asynchronous process: possibly a goroutine for each O we're talking to
- I think in this case it's more valuable to prioritize sql injection protection over disk performance optimization, particularly because these operations are not on the critical path for video processing. This operation seems to be storing external inputs directly into the DB, which poses a real injection risk. I think it's best to use prepared statements in this case, and it's fine to write them one by one as they come in from the async goroutines.

wdyt?

Comment (eladmallel) [reply discovery/db_discovery.go]: Another question I had - where are we capturing the need to augment this caching process with `TicketParam`s?

File: discovery/db_discovery.go (hunk 2 above)
Comment (angyangie) [reply discovery/db_discovery.go]: Definitely on the constant! 

I'm down to try the pattern. Is what you're suggesting adding a context timeout here, like so? `NewDBOrchestratorPoolCache(node *core.LivepeerNode, ctx context.Context)`. In that example, I'm not sure where the `transcodeLoopContext = stubTranscodeLoopContext` gets used. Is `transcodeLoopContext` a global variable that is then overwritten in the tests (in the example)?

File: discovery/db_discovery.go (hunk 3 above)
Comment (angyangie) [reply discovery/db_discovery.go]: Good to starting thinking about `GetOrchestratorInfo` retrieval. For when this happens in the background, it seems like it'd be fine to do async updating. However, when we're setting the cache for the first time on [startup](https://github.com/livepeer/go-livepeer/pull/642/files#diff-205b550ef0ac99b368da16cfb883960bR376), if the number of orchs is very large, wouldn't we experience a significant lag? Could we theoretically have two approaches?

Great idea on multiple INSERTs into a single statement @j0sh. I see the limit is 500 for that, so we'd be able to for now insert all at once. What do you mean by "escape all the input" @j0sh? SQL injections is a valid concern

File: discovery/db_discovery.go (hunk 7 above)
Comment (angyangie) [reply discovery/db_discovery.go:77]: I added the check to `EthOchToDBOrch` to be safe. I'm not sure orchs from `orchestrators, err := node.Eth.RegisteredTranscoders()` can come back nil (i.e. if an error is thrown when `tAddr, err := c.GetFirstTranscoderInPool()` or `tAddr, err = c.GetNextTranscoderInPool(tAddr)` is nil). We can leave both checks in to be safe, though technically as things are written now, if we're checking for a nil orch in this for loop above, we don't need to check in `EthOrchToDBOrch`.

File: common/db_test.go (hunk 8 above)
Comment (angyangie) [reply common/db_test.go]: Repetitive code would be the only downside, as you mentioned, but helper functions sound good. I like this new approach!

File: common/db.go (hunk 9 above)
Comment (angyangie) [reply common/db.go]: Sounds good to me!

File: common/db_test.go (hunk 10 above)
Comment (angyangie) [reply common/db_test.go]: True!

File: common/db_test.go (hunk 11 above)
Comment (angyangie) [reply common/db_test.go]: I like the improvement - you're right in that it does not need to be a specific value

File: common/db_test.go (hunk 12 above)
Comment (angyangie) [reply common/db_test.go]: Yessss!

File: discovery/db_discovery.go (hunk 13 above)
Comment (angyangie) [reply discovery/db_discovery.go]: We don't use this function anywhere else atm, so it can be kept private!

File: discovery/db_discovery.go (hunk 14 above)
Comment (angyangie) [reply discovery/db_discovery.go:48]: Is that a best practice in go? I am fine reorganizing the functions. It does seem like it would improve readability

File: common/db.go (hunk 9 above)
Comment (j0sh) [reply common/db.go]: REPLACE alone does a delete-and-insert which might not necessarily always be what we want. For example, we store a creation timestamp here, which would be lost on a replace. It's not critical to keep that right now, but I can see us wanting to retain other types of fields in the future.

This brings up the fact that the query doesn't seem quite well formed, so thanks for bringing this up. To retain upsert-type behavior we probably need a nested select for the insertion time (and any other fields we wish to preserve). See https://stackoverflow.com/a/4330694

The latest SQLIte does seem to have native upsert as a brand-new feature, but I'm not sure what that means as far as distributing an up-to-date sqlite implementation (IIRC SQLite is typically a system lib but it could well be bundled somewhere in the go-sqlite implementation; can dig into this later). https://www.sqlite.org/draft/lang_UPSERT.html

File: discovery/db_discovery.go (hunk 2 above)
Comment (j0sh) [reply discovery/db_discovery.go]: We don't necessarily need to pass in a context right now.

Correct that transcodeLoopContext is a global, and the stub is set during the testing function that needs it.

In this case, stubbing the ticker out cleanly seems a bit tricky because because the Ticker is a concrete type rather than an interface. I'm hoping we can explicitly set the tick, eg 

```
// default impl
getTicker = func() {
  return time.Ticker(1 * time.Hour)
}
...

// during testing
st := time.NewTicker(1 * time.Hour)
stubTicker = func() {
  return st
}
getTicker = stubTicker
.... do things that depend on the ticker ...
st.C <- time.Now() // generate a tick
... check things that depend on the ticker ...
```
Let me know if that works.

File: discovery/db_discovery.go (hunk 3 above)
Comment (j0sh) [reply discovery/db_discovery.go]: While this would likely be a "low-priority" background task, SQLite writes incur a global lock which blocks all other operations such as reads. We don't want to stall elsewhere in the critical path because of a blocked DB read, so we really should minimize writes as much as possible.

File: common/db_test.go (hunk 8 above)
Comment (j0sh) [reply common/db_test.go]: 😓 I'll go out on a limb here and say that I don't find the "surgical" approach to help readability or understanding the code flow for a few reasons:

* The method names are really long, which takes conscious effort to mentally parse and understand. Even if retaining the surgical approach, I'd really prefer more concise method names, and elaborating on the scenario/expected result as a comment.

* It generates a large amount of boilerplate, which we'd still have with helper functions.

* Separating each test case into its own function, we lose the benefit of code locality. It's harder to quickly scan the tests and reason about the various conditions that are covered (or might not be covered). Here, I was able to quickly see that we were missing nil/empty checks. Also, the progression here can logically be followed: testing empty case, testing one case, testing multiple cases. Splitting that up doesn't improve the readability.

* Related to code locality, one pattern I've used is to check for various conditions following the implementation. This makes it easier to scan the implementation and identify the test case that's covering it. For example, if a function checks and early-exits for conditions A, B, C, we'll have a unit-test checking for A, B, C.  The RPC tests have a bunch of these, eg
https://github.com/livepeer/go-livepeer/blob/master/server/rpc.go#L176 and https://github.com/livepeer/go-livepeer/blob/master/server/rpc_test.go#L147 (admittedly the checks here aren't quite in the proper order; that could be improved.)

Comment (angyangie) [reply common/db_test.go]: I'll try a blended approach and see what y'all think! I see what you mean on making each individual test its own function @j0sh  - that would mean a lot of repeated code. It just doesn't make sense for some of our tests. I am going to try push something that adds comments (which helps with readability), and only breaks up the tests that could stand on their own, minimizing code repetition. The method names are definitely long ... I'll take a stab at using the format @eladmallel suggested, and Josh tell me what you think! Perhaps to make them more readable we can take out the results portion of the name after the last "_". For some of the tests I'm working on, doesn't make too much sense. For now, I'll push something up and open the floor for discussion. We can always edit/adjust after.

It might be good to come up with a "standard" way to write tests, that everyone can be happy with. It would be great to have consistency across all of our test files (though we'd have old tests that wouldn't comply to the new test writing style until we make a deliberate attempt to transition them)

File: common/db.go (hunk 9 above)
Comment (angyangie) [reply common/db.go]: Update - doing some research on how we can make that work. Getting an error that I'm trying to fix locally... 

```db: Unable to fetch orchestrator sql: Scan error on column index 1, name "ethereumAddr": unsupported Scan, storing driver.Value type <nil> into type *string```

File: discovery/db_discovery.go (hunk 2 above)
Comment (angyangie) [reply discovery/db_discovery.go]: Update - Josh and I are working on testing the ticker! `st.C <- time.Now()` does not work the way we thought it did

File: common/db.go (hunk 9 above)
Comment (j0sh) [reply common/db.go]: These changes did the trick for me

```diff
diff --git a/common/db.go b/common/db.go
index 3d7b6b66..9c0a3c91 100644
--- a/common/db.go
+++ b/common/db.go
@@ -56,8 +56,8 @@ var schema = `
 
 	CREATE TABLE IF NOT EXISTS orchestrators (
 		ethereumAddr STRING PRIMARY KEY,
-		createdAt STRING DEFAULT CURRENT_TIMESTAMP,
-		updatedAt STRING DEFAULT CURRENT_TIMESTAMP,
+		createdAt STRING DEFAULT CURRENT_TIMESTAMP NOT NULL,
+		updatedAt STRING DEFAULT CURRENT_TIMESTAMP NOT NULL,
 		serviceURI STRING
 	);
 
@@ -145,8 +145,7 @@ func InitDB(dbPath string) (*DB, error) {
 	d.selectOrchs = stmt
 
 	// updateOrchestrators statement
-	stmt, err = db.Prepare("INSERT OR REPLACE INTO orchestrators(updatedAt, serviceURI, ethereumAddr) VALUES(datetime(), ?, ?)")
-	// stmt, err = db.Prepare("INSERT OR REPLACE INTO orchestrators(updatedAt, serviceURI, ethereumAddr) VALUES(datetime(), ?, (SELECT ethereumAddr FROM orchestrators WHERE ethereumAddr = ?))")
+	stmt, err = db.Prepare("INSERT OR REPLACE INTO orchestrators(updatedAt, serviceURI, ethereumAddr, createdAt) VALUES(datetime(), ?1, ?2, (SELECT createdAt FROM orchestrators WHERE ethereumAddr = ?2))")
 	if err != nil {
 		glog.Error("Unable to prepare updateOrchestrators stmt ", err)
 		d.Close()
```

Comment (j0sh) [reply common/db.go]: >  I'm not sure what that means as far as distributing an up-to-date sqlite implementation 

Looks like we're in luck and go-sqlite **does** incorporate [its own copy of sqlite](https://github.com/mattn/go-sqlite3/commit/554bc0fb2e0ee7ec0105f4296b7fa5e01f85aa32), and the repo looks to be fairly up-to-date; we just need to refresh the goclient's vendored copy. This also partially explains why the goclient takes so long to build/link...

Doesn't to be done in this PR though; only if we really want to use the new upsert syntax here.

File: discovery/db_discovery.go [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+
+	ticker := getTicker()
+	go func(node *core.LivepeerNode) {
+		for {
+			select {
+			case <-ticker.C:
+				orchestrators, err := node.Eth.RegisteredTranscoders()
+				if err != nil {
+					glog.Error("Could not refresh DB list of orchestrators: ", err)
+					return
```
Comment (j0sh) [inline discovery/db_discovery.go]: Probably best to `continue` here; the issue may have resolved itself by the next tick

File: discovery/db_discovery.go [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+				}
+				_, dbOrchErr := cacheDBOrchs(node, orchestrators)
+				if dbOrchErr != nil {
+					glog.Error("Could not refresh DB list of orchestrators: cacheDBOrchs err")
+					return
```
Comment (j0sh) [inline discovery/db_discovery.go]: Ditto on `continue`

File: discovery/db_discovery.go [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+	}
+
+	for _, orch := range orchs {
+		if orch == nil {
+			continue
+		}
+		dbOrch := ethOrchToDBOrch(orch)
+		if err := node.Database.UpdateOrch(dbOrch); err != nil {
+			glog.Error("Error updating Orchestrator in DB: ", err)
+			return nil, err
```
Comment (j0sh) [inline discovery/db_discovery.go]: `continue`

File: common/db.go (hunk 9 above)
Comment (angyangie) [reply common/db.go]: 🙀 thank you @j0sh! Adding `createdAt` and numbering the ?'s is genius. Why is it better to select on`createdAt` here instead of `ethereumAddr`? 
```sql
(SELECT createdAt FROM orchestrators WHERE ethereumAddr = ?2))")

Comment (j0sh) [reply common/db.go]: Because REPLACE is a delete-and-insert, we need to preserve columns that aren't explicitly inserted. The only column we don't insert here is `createdAt`, so we select that based on the `ethereumAddr`.

If the row doesn't exist (eg, this is an insert) then the select returns NULL. With NULL, the default value (CURRENT_TIMESTAMP) ultimately gets set for the column. This is why the `NOT NULL` constraint had to be set above; it was also added to the updatedAt for completeness.

File: discovery/db_discovery.go (hunk 3 above)
Comment (eladmallel) [reply discovery/db_discovery.go]: Indeed writing to the DB involves locking, which is a certain cost. There is also the cost of writing more complex INSERT code, testing it and maintaining it.

What's your framework for deciding which performance hits (such as DB locks) are worth having suboptimal code, versus hits that aren't worth it?

File: common/db.go (hunk 9 above)
Comment (angyangie) [reply common/db.go]: Makes sense. Thank you!

File: common/db_test.go (hunk 8 above)
Comment (eladmallel) [reply common/db_test.go]: Seems like this is a harder point to achieve consensus around, since I actually find the long test functions with comments along the way harder to parse than the shorter functions I've been trying out.

Since I find consistency very important, I would love for us to land on a specific style all of us commit to.

@j0sh as the leader of this repo I believe you should drive this decision. I realize it is probably going to be long test functions, which will give me the opportunity to improve my code parsing skills :D

File: discovery/db_discovery.go (hunk 14 above)
Comment (eladmallel) [reply discovery/db_discovery.go:48]: I honestly don't know the idiomatic Go file structure. Found [this thread](https://www.reddit.com/r/golang/comments/2xn1ws/idiomatic_go_ordering_of_functions_in_source_files/) with some thoughts, in which no "official" position was shared.

I very much connect with doing our best to make code read coherently from top to bottom, which is why I like placing public functions before private ones, so that when you reach the private function you already know in what context it's used. I also find it very helpful to make sure all of a type's functions are batched together (they can be in separate files if it makes sense, but they shouldn't be scattered across a file with other functions intermingled). 

Perhaps @j0sh should weigh in here and suggest a standard we should stick to?

Comment (j0sh) [reply discovery/db_discovery.go:48]: I don't really have a strong preference about function placement. The proposed approach sounds good.

Personally, I typically start off with keeping function definitions as close as possible to its use to make quick scanning (or even editing) easy, even if this means some interleaving of methods and functions. Of course, there are lots of exceptions to this, and code tends to drift anyway without a conscious effort to keep things organized.

The C programmer in me has always been slightly bemused about defining a function *after* it's used (without a forward declaration) -- it feels a bit like using a variable before it's declared. Lots of languages have this feature though, so why not.

File: discovery/db_discovery.go [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+
+func NewDBOrchestratorPoolCache(node *core.LivepeerNode) *DBOrchestratorPoolCache {
+	if node.Eth == nil {
+		glog.Error("Could not refresh DB list of orchestrators: LivepeerNode nil")
+		return nil
+	}
+
+	ticker := getTicker()
+	go func(node *core.LivepeerNode) {
+		for {
```
Comment (yondonfu) [inline discovery/db_discovery.go]: I believe an alternative to nesting the single case select statement within this for loop is to just use a for loop and range over the `ticket.C` channel:

```
for _ := range ticker.C {
    ...
}
```

File: discovery/db_discovery.go [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+
+	var uris []string
+	for _, orch := range orchs {
+		uri := orch.ServiceURI
+		uris = append(uris, uri)
+	}
+
+	orchPool := NewOrchestratorPool(dbo.node, uris)
+
+	orchInfo, err := orchPool.GetOrchestrators(numOrchestrators)
```
Comment (yondonfu) [inline discovery/db_discovery.go]: Should `orchInfo` be plural if it is an array?

Comment (angyangie) [reply discovery/db_discovery.go]: Yes! I also changed `uri` in the `orchestratorPool` struct to `uris`, which makes more sense as well

**Comment by angyangie** [comment]
@j0sh @eladmallel @yondonfu, what do you think of the changes? Any other comments?

**Review by j0sh — APPROVED** [review]
LGTM as-is.

Once rebased on top of master, please give it another run-through of manual tests and ensure that e2e transcoding works (well, up to [segment submission](https://github.com/livepeer/go-livepeer/issues/666)), and that the DB is updated periodically.

**Comment by angyangie** [comment]
Thanks @j0sh. Will do!

**Review by yondonfu — APPROVED** [review]
